packages/datadog_grpc_interceptor/lib/src/datadog_grpc_interceptor.dart

```diff
@@ -0,0 +1,102 @@
+// Copyright 2019-Present Datadog, Inc.
+
+import 'dart:io';
+
+import 'package:datadog_flutter_plugin/datadog_flutter_plugin.dart';
+import 'package:datadog_grpc_interceptor/src/tracing_util.dart';
+import 'package:grpc/grpc.dart';
+import 'package:uuid/uuid.dart';
+
+class DatadogTracingHeaders {
+  static const traceId = 'x-datadog-trace-id';
+  static const parentId = 'x-datadog-parent-id';
+
+  static const sampled = 'x-datadog-sampled';
+  static const samplingPriority = 'x-datadog-sampling-priority';
+  static const origin = 'x-datadog-origin';
+}
+
+// TODO: These are defined in the main datadog_flutter_plugin package but aren't
+// public. Probably want to fix that.
+class DatadogPlatformAttributeKey {
+  /// Trace ID. Used in RUM resources created by automatic resource tracking.
+  /// Expects `String` value.
+  static const traceID = '_dd.trace_id';
+
+  /// Span ID. Used in RUM resources created by automatic resource tracking.
+  /// Expects `String` value.
+  static const spanID = '_dd.span_id';
+}
+
+/// A client GrpcInterceptor which enables automatic resource tracking and
+/// distributed tracing
+///
+/// Note: This only supports intercepting unary calls. It does not intercepting
+/// streaming calls.
+class DatadogGrpcInterceptor extends ClientInterceptor {
+  final Uuid uuid = const Uuid();
+  final DatadogSdk _datadog;
+  final ClientChannel _channel;
+
+  DatadogGrpcInterceptor(
+    this._datadog,
+    this._channel,
+  );
+
+  @override
+  ResponseFuture<R> interceptUnary<Q, R>(ClientMethod<Q, R> method, Q request,
+      CallOptions options, ClientUnaryInvoker<Q, R> invoker) {
+    final host = _channel.host;
+    final path = method.path;
+    final String fullPath;
+    if (host is InternetAddress) {
+      fullPath = '${host.host}:${_channel.port}$path';
+    } else {
+      fullPath = '$host:${_channel.port}$path';
+    }
+
+    bool shouldAppendTraces = _datadog.traces != null &&
+        _datadog.isFirstPartyHost(Uri.parse(fullPath));
+
+    String? traceId;
+    String? parentId;
+    if (shouldAppendTraces) {
+      traceId = generateTraceId();
+      parentId = generateTraceId();
+    }
+
+    final rumKey = uuid.v1();
+    _datadog.rum?.startResourceLoading(
+      rumKey,
+      RumHttpMethod.get,
+      fullPath,
+      {
+        "grpc.method": method.path,
+        if (shouldAppendTraces) ...{
+          DatadogPlatformAttributeKey.traceID: traceId,
+          DatadogPlatformAttributeKey.spanID: parentId
+        }
+      },
+    );
+
+    if (shouldAppendTraces) {
+      options = options.mergedWith(CallOptions(metadata: {
+        DatadogTracingHeaders.origin: 'rum',
+        DatadogTracingHeaders.samplingPriority: '1',
+        DatadogTracingHeaders.traceId: traceId!,
+        DatadogTracingHeaders.parentId: parentId!,
+      }));
+    }
+
+    final future = invoker(method, request, options);
+    future.then((v) {
+      _datadog.rum?.stopResourceLoading(rumKey, 200, RumResourceType.native);
```

Comment:
that is actually a question to think about: should we use `native` here or not. In RN we use `xhr` type (despite the fact that HTTP transport in RN is backed by native libs anyway).

Reply:
Another good question. When should we use xhr vs. native? To be honest, I think I used xhr in some places in the tracking_http_client.

Reply:
It was all `xhr` until not so long time ago: last September we added `native` type (https://github.com/DataDog/dd-sdk-android/pull/702 and https://github.com/DataDog/dd-sdk-ios/pull/605) to differentiate the calls made from native side vs cross-platform.

RN is using `xhr`, because calls made there are essentially backed by `XMLHttpRequest` (which is just an interface, and HTTP stack is implemented by OkHttp anyway on Android, for example).

With Flutter we may still have native modules in the user app, which will do `native` calls, but then the question is how do we categorize Flutter calls.

Reply:
I'll leave it as native for now. I think that's the most accurate?

Reply:
Having `native` is fine imo for now. But let's also have a TODO item for a further discussion, if we want to separate native requests from flutter requests.